### PR TITLE
Add a babel-macro

### PR DESCRIPTION
**Pull request by lucleray**
Since https://github.com/styled-components/styled-components/pull/1256 seems to be not active anymore, I created a new PR to add babel-macro to styled-components.

- [x] Macro
- [x] Import visitors from `babel-plugin-styled-components`
- [x] Tests
- [x] Flow types (enough for flow check to pass)
- [x] Typescript types
- [x] Export as `styled-components/macro`
- [x] Implement visitors from `babel-plugin-styled-components`

### CODE REVIEW

**Comment by probablyup** [comment]
Let's add a test case to verify that all helpers exported by `src/index.js` have a matching implementation in the macro so if we add or change something in the future, we'll be notified via tests if we're missing something.

**Comment by lucleray** [comment]
@probablyup 
> Let's add a test case to verify that all helpers exported by `src/index.js` have a matching implementation in the macro so if we add or change something in the future, we'll be notified via tests if we're missing something.

I've added a test for that here : https://github.com/styled-components/styled-components/blob/7992c61d99cf4ce9f665ca0538b67643534d47ac/src/macro/test/macro.test.js#L64-L70

**Comment by lucleray** [comment]
With the current implementation, that's the way it works :

```jsx
import styled, { css, keyframes, injectGlobal, ThemeProvider } from 'styled-components/macro'

// do things with imports
css`...`
keyframes`...`
injectGlobal`...`
styled.div`...`
<ThemeProvider />
```

It will get transpiled to something like this (just focusing on what the macro does, not the other babel plugins) :
```jsx
import _styled, { _css, _keyframes, _injectGlobal, _ThemeProvider } from 'styled-components'

_css() 
_keyframes()
// lighter precompiled versions

_injectGlobal`...`
// not modified because babel-plugin-styled-components keeps it like this

_styled.div.withConfig({ displayName: '' })()
// adds withConfig (displayName, ...) and use the lighter version

<_ThemeProvider /> 
// kept like this, just moved the import from styled-components/macro to styled-components
```

**Comment by lucleray** [comment]
@probablyup 
Could you let me know if the current behaviour of the macro is good on your side ?

I'll wait for your input to add types 🙂

**Comment by lucleray** [comment]
I added just enough flow types for `flow check` to pass. I'm not sure we need more ?

If someone knows Flow, I'd love some advice on the flow types in this PR 😬

**Comment by probablyup** [comment]
@lucleray could you provide me with a testing guide? I don't use CRA so not sure how to go about making sure this works

**Comment by lucleray** [comment]
@probablyup 

I created an example repo here : https://github.com/lucleray/cra-styled-components

First, you need to set it up.

1. Clone this PR, `yarn install`, `yarn build` and `yarn link`
2. Clone [the example repo](https://github.com/lucleray/cra-styled-components), `yarn install` and `yarn link styled-components`

Once this is set up, you can test if the macro does the right thing.

I added two styled components in `src/App.js` : https://github.com/lucleray/cra-styled-components/blob/4883352b32a9adfbcfad02710b79152a7007deae/src/App.js#L4-L19

Run cra :
- To run cra in dev mode : `yarn start`
- To run cra build : `yarn build`

Then you can check that the code is correctly transpiled :
- In dev mode, I use the console's sources tab
- When it's build, I look in the `build/` directory

I look for the main chunk (`static/js/main.<...>.chunk.js`) and search for `StyledDiv` in the code. I see something like this :

```js
var StyledDiv = styled_components__WEBPACK_IMPORTED_MODULE_5__["default"].div.withConfig({
  displayName: "App__StyledDiv",
  componentId: "sc-19hhnfk-0"
})(["background:red;padding:5px;margin-top:1em;border-radius:5px;"]);
```

Similarly, I search for `StyledNoMacroDiv` :

```js
var StyledNoMacroDiv = styled_components__WEBPACK_IMPORTED_MODULE_6__["default"].div(_templateObject());
```

**Comment by lucleray** [comment]
The example repo was quite easy to setup :

```
npx create-react-app@next --scripts-version=@next cra-styled-components
cd cra-styled-components
yarn add styled-components
```

And this is all you need to do to start using CRA + styled-components once this macro will be added to styled-components 🔥

File: src/macro/index.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+        if (!templatePath) {
+          throw new MacroError(
+            `Invalid usage : ${refName}. ${refName} needs to be used with a template literal, like this : ${refName}\`background: red;\`.`
+          )
+        } else {
+          // merge config into the state
+          const stateWithOpts = { ...state, opts: config }
+          // run babel-plugin-styled-components appropriate visitors
+          minify(t)(templatePath, stateWithOpts)
```
Comment (probablyup) [inline src/macro/index.js]: Is there a way to reuse the setup from the babel plugin wholesale instead of having to recompose all of these? I can't see us maintaining this properly if it isn't more automatic. It's also a bit more complicated as of 1.8 since we're handling call expressions too

Comment (lucleray) [reply src/macro/index.js]: That sounds like a good idea 😀

**Comment by lucleray** [comment]
@probablyup 
I've updated according to the comments on the PR.

The macro is more simple codewise now :
1. it takes files where `styled-components/macro` is imported
2. it replaces `styled-components/macro` by `styled-components` and updates references
3. it applies `babel-plugin-styled-components` to the file

So whenever `babel-plugin-styled-components` is updated, things will just work 👍

**Comment by lucleray** [comment]
I also rebased onto `develop` and changed the base branch to `develop`, so you'll need to reset your local branch with the changes.

**Comment by alexandernanberg** [comment]
Just throwing this out there but would it make more sense to have to macro in `babel-plugin-styled-components` and then just have it as a dependency here? I think that's how emotion have it setup 🤔

**Comment by lucleray** [comment]
@alexanderanberg 
I think it's better to put it in styled-components.

The `import styled from 'styled-components/macro'` is replaced by `import styled from 'styled-components'` so you install only 1 dependency and things work.

**Comment by strayiker** [comment]
@lucleray @alexandernanberg shouldn't be code of `macro` be placed into `babel-plugin-styled-components` and then re-exported from `styled-components`?

**Comment by lucleray** [comment]
@strayiker I'm not sure why you want to put the code that we need in `styled-components` in `babel-plugin-styled-components` ? 🤔

The macro just imports the plugin and uses it through the same api like when it's imported as a plugin. So if there's any changes in `babel-plugin-styled-components`, the macro will work just fine.

**Comment by strayiker** [comment]
@lucleray `Macro` is still the babel plugin, although it just imports the plugin and working slightly different way, it's still the plugin for babel. Because of that i think it should be placed inside the `babel-plugin-styled-components`. May be i'm wrong, ofc.

File: src/macro/index.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
+// @flow
+import { createMacro, MacroError } from 'babel-plugin-macros';
+import babelPlugin from 'babel-plugin-styled-components';
+
+const allowedImports = [
```
[the remark this answers is not in the document]
Comment (lucleray) [reply src/macro/index.js]: Agree!

This test is also not necessary once I've updated the PR with this so I'll remove it : https://github.com/styled-components/styled-components/blob/a4188f0524f1f7b24c674ec27c8e1dc9b932a3c9/src/macro/test/macro.test.js#L64-L70

File: package.json [hunk 3]
```diff
@@ -63,6 +63,7 @@
   "homepage": "https://styled-components.com",
   "dependencies": {
     "@emotion/is-prop-valid": "^0.6.8",
+    "babel-plugin-styled-components": "^1.8.0",
```
[the remark this answers is not in the document]
Comment (lucleray) [reply package.json]: If it becomes a peer dependency, then we lose the simplicity of setting up styled-components with create-react-app :
1. `yarn add styled-components`
2. `import styled from 'styled-components/macro'`

**Comment by strayiker** [comment]
@lucleray This will make it easier to find and improve code in the future. Whenever you change the code of the full featured babel-plugin, you probably always be aware of that the `macro` is not broken by this changes. If it placed in the same repo, you will see the unit tests is failed. Does it make sense?

**Comment by lucleray** [comment]
@strayiker @mxstbr 
Ok! I'll keep the code here.

Waiting for @probablyup's answer :
https://github.com/styled-components/styled-components/pull/2032#discussion_r222711468  : "should babel-plugin-styled-components dependency be a peerDependency or a dependency ?" 🙂

**Comment by probablyup** [comment]
Yes it should be a peer dep

**Comment by lucleray** [comment]
@probablyup 
> Yes it should be a peer dep

What is the reason for that ?

Emotion is listing it as a dependency : https://github.com/emotion-js/emotion/blob/f8e795fe77d601b718ea4da4a906f820599c20a5/packages/css/package.json#L12

**Comment by probablyup** [comment]
So it dedupes properly if another third party library adds it

**Comment by lucleray** [comment]
@probablyup 
Ok, I moved it to peerDependency here : https://github.com/styled-components/styled-components/pull/2032/commits/5cef31124c50b5b1702a303879e4153bc62afa84

I'm still not sure the dedup is more important than the ease of use/simplicity though 🤔

**Comment by mxstbr** [comment]
> I'm still not sure the dedup is more important than the ease of use/simplicity though 🤔

It won't be obvious for create-react-app users that they have to `import styled from 'styled-components/macro'` anyway, which means we will need to add this to the documentation (once as its own item "Usage with create-react-app" and once in the API section somehow) anyway—so we might as well add a second step to the documentation that says that you first have to install the Babel plugin.

In fact, this might be a compelling reason to move the macro to the Babel plugin instead after all (sorry 🙈). Imagine the docs as it is right now:

```
## Usage with create-react-app

First, install the Babel plugin:

$ npm i babel-plugin-styled-components

Then use the Babel macro:

import styled from 'styled-components/macro';
```

It's pretty unobvious that you need to install the plugin alongside before being able to use the macro mode. Contrast with this explanation:

```
## Usage with create-react-app

First, install the Babel plugin:

$ npm i babel-plugin-styled-components

Then use the Babel macro:

import styled from 'babel-plugin-styled-components/macro';
```

That's more obvious, if a bit more typing in the import statement.

Not sure what's better or worse, what do you think?

**Comment by lucleray** [comment]
@mxstbr 

IMO, the best way experience for users is to just install styled-components and import it through the macro and it works :
```
## Usage with create-react-app

Install styled-components:

$ npm i styled-components

Use it through the macro:

import styled from 'styled-components/macro';
```

You don't need to install `babel-plugin-styled-components` at all ! **(That means babel-plugin-styled-components should be a dep and not a peer dep)**

One of the reason for babel-plugin-macros to exist as a replacement of babel plugins is (listed in [this article](https://babeljs.io/blog/2017/09/11/zero-config-with-babel-macros)) :
> [Babel plugins] can lead to confusion because when looking at code in a project, you might not know that there's a plugin transforming that code.

When you need to install babel-plugin-styled-components as a user  :
- if you `import styled from 'styled-components/macro'` then it's not explicit why you need babel-plugin-styled-components because it's not in your source code
- if you `import styled from 'babel-plugin-styled-components/macro'` then it's not clear why you need styled-components (if this is what we want to do, then why not use a dedicated package `styled-components.macro` like [this one](https://github.com/lucleray/styled-components.macro))

Also, other libraries are doing the same :
- styled-jsx => `import css, { resolve } from 'styled-jsx/macro'`, no need to install `babel-plugin-syntax-jsx`
- emotion => `import { css } from 'emotion/macro'` , no need to install `babel-plugin-emotion`.

To sum up, this is what I think is best **for users** :

|| description | how it works |
|-|-|-|
| Solution 1 | babel-plugin-styled-components is a dep (not a peer dep) | `import styled from 'styled-components/macro'` |
| Solution 2 | create a dedicated package `styled-components.macro` (maybe move [this one](https://github.com/lucleray/styled-components.macro) to the styled-components org)  | `import styled from 'styled-components.macro'` |

**Comment by mxstbr** [comment]
I see, good thinking. I think adding it as a dep is really the best UX.

@probablyup let's put it as a dep with a really lenient version, maybe `=> 1.0`? People should be using it anyway so I don't feel bad about installing it alongside automatically.

**Comment by gilbsgilbs** [comment]
> It won't be obvious for create-react-app users that they have to import styled from 'styled-components/macro' anyway

Speaking of this, what will happen if they don't? Will it still work? Work but silently not apply transforms? Throw some error? Is there any error / linting rule one can enable to ensure the macro is consistently imported in CRA instead of the "normal" import? Can we document this if it's not already?

**Comment by mxstbr** [comment]
> Speaking of this, what will happen if they don't? Will it still work? Work but silently not apply transforms?

Just like it does today it will work as normally, just without any Babel transforms.

> Can we document this if it's not already?

See my reply further up, 

> which means we will need to add this to the documentation (once as its own item "Usage with create-react-app" and once in the API section somehow) anyway

**Comment by lucleray** [comment]
> Speaking of this, what will happen if they don't? Will it still work? Work but silently not apply transforms?

> Just like it does today it will work as normally, just without any Babel transforms.

Yes, this is true currently, unless there is a file with 2 imports like this :
```js
import styled from 'styled-components'
import styled2 from 'styled-components/macro'
// ...
```

In that case, both styled imports will be transformed. 

This is because of the current implementation. When `import from 'styled-components/macro'` is present in the file, the macro will pass the file to the babel plugin and the babel plugin will transform the whole file, not just locally.

The implementation is like this because we want the macro to be as close to the babel plugin as possible and easy to maintain when the babel plugin is updated. *If we wanted to just transform locally, the macro should be aware of **styled-components' API** and that would be hard to maintain.*

For example, if it is used like this :
```js
import styled from 'styled-components'
import styled2 from 'styled-components/macro'

styled2.div`background: red;`
```

If we want to only transform locally, we need to go up twice the tree to apply the transformation to the thing that matters to styled components, the `TaggedTemplateExpression ` :
```
styled2 ===(parentPath)===> styled2.div ===(parentPath)===> styled2.div`background: red;`
```

But this is specific to the default import, so we need to do the same for the other imports (css, keyframes, ...). *So we start to implement **styled-components' API** in the macro itself.* That's why the current implementation just pass the whole file to babel plugin and doesn't do things locally.

AFAIK, I can't find a real use case for importing both `styled-components` and `styled-components/macro` in the same file, so I think this behaviour is acceptable. What do you think ?

**Comment by mxstbr** [comment]
> AFAIK, I can't find a real use case for importing both styled-components and styled-components/macro in the same file, so I think this behaviour is acceptable. What do you think ?

Yeah that's totally fine.

**Comment by probablyup** [comment]
I wish there was a way to have the macro applied automatically without having to specify a different import path...

**Comment by probablyup** [comment]
Maybe in the docs for this we should also recommend setting up aliasing so "styled-components" -> "styled-components/macro"

**Comment by probablyup** [comment]
Also it looks like the community convention seems to be `name.macro` rather than `name/macro`. Will the former work so people don't get confused?

**Comment by probablyup** [comment]
Alright it's going in! Thanks for all the amazing work here @lucleray!

**Comment by mxstbr** [comment]
Thank you so much for helping us improve styled-components! Based on our [Community Guidelines](https://github.com/styled-components/styled-components/blob/master/CONTRIBUTING.md) every person that has a PR of any kind merged is offered **an invitation to the styled-components organization**—that is you right now!

Should you accept, you'll get write access to the main repository. (and a fancy `styled-components` logo on your GitHub profile!) You'll be able to label and close issues, merge pull requests, create new branches, etc. We want you to help us out with those things, so don't be scared to do them! Make sure to read our [contribution and community guidelines](https://github.com/styled-components/styled-components/blob/master/CONTRIBUTING.md), which explains all of this in more detail. Of course, if you have any questions just let me know!

**Comment by gilbsgilbs** [comment]
Thanks for the work on this PR.

> I wish there was a way to have the macro applied automatically without having to specify a different import path...

> Maybe in the docs for this we should also recommend setting up aliasing so "styled-components" -> "styled-components/macro"

It's not aliasing, but [`no-restricted-imports`](https://eslint.org/docs/rules/no-restricted-imports) can at least prevent mistakes. Does this miss any import?

```yaml
rules:
  no-restricted-imports:
    - error
    - paths:
      - name: 'styled-components'
        message: 'Please import from styled-components/macro.'
      patterns:
        - '!styled-components/macro'
```


I find it more explicit and intended than aliasing, however I'd be interested in knowing how you would implement aliasing. Using [resolve.alias](https://webpack.js.org/configuration/resolve/#resolve-alias)?

**Comment by lucleray** [comment]
Great to see this merged 👌

@gilbsgilbs 
If we want to add a warning/aliasing, it should only be for people who are using babel-plugin-macros, and that seems very tricky to detect, so I'm not sure that would work 🤔
